common/argparse.h

```diff
@@ -0,0 +1,870 @@
+                 llvm::function_ref<void(FlagBuilder&)> build);
+    void AddIntegerOption(const ArgInfo& info,
+                          llvm::function_ref<void(IntegerArgBuilder&)> build);
+    void AddStringOption(const ArgInfo& info,
+                         llvm::function_ref<void(StringArgBuilder&)> build);
+    void AddOneOfOption(const ArgInfo& info,
+                        llvm::function_ref<void(OneOfArgBuilder&)> build);
+    void AddMetaActionOption(const ArgInfo& info,
+                             llvm::function_ref<void(ArgBuilder&)> build);
+
+    void AddIntegerPositionalArg(
+        const ArgInfo& info,
+        llvm::function_ref<void(IntegerArgBuilder&)> build);
+    void AddStringPositionalArg(
+        const ArgInfo& info, llvm::function_ref<void(StringArgBuilder&)> build);
+    void AddOneOfPositionalArg(
+        const ArgInfo& info, llvm::function_ref<void(OneOfArgBuilder&)> build);
+
+    void AddSubcommand(const CommandInfo& info,
+                       llvm::function_ref<void(CommandBuilder&)> build);
+
+    // Subcommands can be hidden from the help listing of their parents with
+    // this setting. Hiding a subcommand doesn't disable its own help, it just
+    // removes it from the listing.
+    void HelpHidden(bool is_help_hidden);
+
+    // Exactly one of these three should be called to select and configure the
+    // kind of the built command.
+    void RequiresSubcommand();
+    void Do(ActionT action);
+    void Meta(ActionT meta_action);
+
+   private:
+    friend Args;
+
+    explicit CommandBuilder(Command& command, MetaPrinter& meta_printer);
+
+    auto AddArgImpl(const ArgInfo& info, ArgKind kind) -> Arg&;
+    void AddPositionalArgImpl(const ArgInfo& info, ArgKind kind,
+                              llvm::function_ref<void(Arg&)> build);
+    void Finalize();
+
+    Command& command_;
+    MetaPrinter& meta_printer_;
+
+    llvm::SmallDenseSet<llvm::StringRef> arg_names;
+    llvm::SmallDenseSet<llvm::StringRef> subcommand_names;
+  };
+
+  // Builds a description of a command and then parses the provided arguments
+  // for that command.
+  //
+  // This is the main entry point to both build up the description of the
+  // command whose arguments are being parsed and to do the parsing. Everything
+  // is done in a single invocation as the common case is to build a command
+  // description, parse the arguments once, and then run with that
+  // configuration.
+  //
+  // The `out` stream is treated like `stdout` would be for a Unix-style command
+  // line tool, and `errors` like `stderr`: any errors or diagnostic information
+  // are printed to `errors`, but meta-actions like printing a command's help go
+  // to `out`.
+  static auto Parse(llvm::ArrayRef<llvm::StringRef> unparsed_args,
+                    llvm::raw_ostream& out, llvm::raw_ostream& errors,
+                    const CommandInfo& command_info,
+                    llvm::function_ref<void(CommandBuilder&)> build)
```

Comment:
What's the advantage you intend to gain from putting the configuration logic in this `build` function? I know you get repeatability, but this library is optimized for one-shot configuration and parsing. What was the thinking there?

Reply:
I guess, where would the configuration logic go instead?

I need some builder-like logic here. A fluent-style API looks really nice, but because these can *recurse* in complex ways, a lambda-passed-a-builder seemed a bit more clean.

If there are other ways to get the builder / DSL of constructing these into the API, happy to explore them.

Reply:
I've thought about it a bit. I prefer this lambda design over the fluent style for a couple of reasons:

a) More control over return types on the config functions. Fluent-style APIs are obligated to return the mutable ref to themselves.

b) Better clarity in the use of `CommandBuilder`. Often the fluent-style objects end up with comments in the header indicating when it is OK to call the configuration functions. For example, it might be possible to call additional config functions after parsing, and the compiler does not detect this misuse without cleverness.

c) Cleaner code. Those long call chains can lead to difficult-to-read (and even harder-to-debug) code stanzas.

d) Good configuration API separation from the library's other parts.

The *only* downside I see is that this promotes centralization of the configuration code. Without call chaining, all library configuration must happen in one function in one file. This could be a feature, but I imagine over time Carbon will grow in complexity, and the `driver.cpp` file will grow along with it.

Again, not saying this is the wrong approach. I think the "Where would the config logic go instead?" question looms large here. Some flags will likely impact multiple subsystems, so putting the config logic with the subsystem feels like a poor separation of concerns.

So I think my feedback is "LGTM so long as we accept a really long client config file."

Reply:
I think I note this elsewhere, but I find the lambda approach very difficult to read. e.g., consider the test example:

```
  auto parse = [&](llvm::ArrayRef<llvm::StringRef> args) {
    return Args::Parse(
        args, llvm::errs(), llvm::errs(), TestCommandInfo, [&](auto& b) {
          b.AddFlag({.name = "flag"}, [&](auto& arg_b) { arg_b.Set(&flag); });
          b.AddIntegerOption({.name = "option1"},
                             [&](auto& arg_b) { arg_b.Set(&integer_option); });
          b.AddStringOption({.name = "option2"},
                            [&](auto& arg_b) { arg_b.Set(&string_option); });
          b.Do([] {});
        });
  };
  ```

From looking at the toolchain, I think that accurately reflects what real code is intended to look like.

I think that could have been something more like:

```
ArgParser parser(llvm::errs(), llvm::errs(), TestCommandInfo);
parser.AddFlag("flag", "desc", &flag);
parser.AddIntegerOption("option1", "desc", &integer_option);
parser.AddStringOption("option2", "desc", &string_option);
parser.Parse(args);
```

I might've commented on this elsewhere, as I'd been expecting an API closer to argparse in structure. I'm commenting here for context given lnihlen's related response, but I don't see why this is just lambda versus fluent style builder.

I understand that extra complexity would need to be added in spots where custom flag parsing is needed, but I think the latter approach would still be quite adaptable. To me, the difference is that it remains simple for simple use-cases, which I think will remain the majority.

Reply:
(we also chatted a bit about this in person, but trying to relay the core of it here)

I'm happy to try to iterate on the DSL used to describe commands FWIW. But I did try something similar to what you outline, and wasn't able to arrive at a solution that really worked well. Not saying it's impossible, but the result seemed wasteful or difficult to read, or both (or didn't let me implement help .... there were a lot of experiments that I backed out).

The most pressing challenge I encountered for a method-setting API (as opposed to a fluent builder or a lambda builder) was similar to the problem with the fluent builder. I struggled to build a general way to nest and recurse to build subcommands in a way that was general and not baking in a specific subcommand structure. I think the subcommands are where the lambda building really shines. The other place where I find it nice is the `OneOf` because we actually again need to recurse a tiny bit to build the individual values.

Between the two valuable spots, it seemed marginally better to be consistent and get used to the lambda-builder pattern than to more pervasively switch back and forth between lambda-builder and some other pattern (whether fluent-builder or method-setting). But I can definitely explore having a bit more of a hybrid.

One option that @zygoloid had suggested earlier when chatting about this before I even had the PR ready was to specialize the case where we call exactly one method in the lambda-builder so we don't need a lambda at all there. I think I see an easy way to do that. WDYT? Also, would you like to see that in this PR or in a follow-up?

Reply:
We can iterate on the API.